### ISSUE TITLE
Allow rules to start using another lexer and let it pop back to the main lexer

### ISSUE THREAD

**Issue by jeanas** [pygments/pygments#2409]
Currently, lexers such as HtmlLexer that embed content in other languages usually make use of `using()`, which requires detecting in advance where the inner content is going to end. For some cases it works well, for some others it's more fragile. As an example, a string containing `</script>` in JS code will confuse Pygments.

We also have DelegatingLexer, which similarly requires the subclass to reimplement enough logic from the base class to detect the end of the embedded content correctly.

If the content is not ended by a marker but by ending the outer paren/bracket in a set of nested parens/brackets, e.g. in Flex, Bison or LilyPond, this is impractical, as regexes can't detect delimiter nesting. Our current solution is subclassing (see LilyPondLexer for an example), but this is a bit convoluted. The base class needs to delegate its root state to an auxiliary state, as the subclass overrides it.

I think it would be interesting to be able to write rules like

```
(r'regex', Token.Foo, embedded(SchemeLexer, state='value')),
```

where `embedded` (WANTED: better name) delegates to the other lexer, and when the other lexer's stack becomes empty again, instead of restarting in the root state, it gives back control to the main lexer.

WDYT?

**Comment by birkenfeld** [comment]
If it's useful in some instances, why not?

The only caveat I see is that states then become "public API" (in quotes since it's still internal to pygments, although `embedded` could then also be used by externally added lexers), and this needs to be taken into account when refactoring the embedded lexer.

It also requires states to be properly scoped, but this can be done on a case by case basis.

**Comment by Anteru** [comment]
I think this makes sense, but the question is if `DelegatingLexer` could be augmented to handle this? It feels like there's going to be two ways to do similar things here (`embedded` vs `DelegatingLexer`) and I'd rather have only one canonical way to do it. Given you already mentioned delegating lexers, it's quite obvious that's the concept we want here, but it seems like the original implementation is more complicated than necessary.

**Comment by birkenfeld** [comment]
> Re public API: yes, true. On the other hand, this is already the case with using(X, state=blabla) and subclassing.

Oh right, I forgot we have this. Then it's already moot.

> Re scoping: not sure I understand, could you rephrase?

Forget about it, it was a not very useful thought expressed badly.

> it's quite obvious that's the concept we want here, but it seems like the original implementation is more complicated than necessary.

Yes, it would be interesting if we could hit both flies with one swatter. :smile:
